docs/adr/007_cups_firmware_upgrade.md

```diff
@@ -0,0 +1,211 @@
+>general CUPS protocol implementation please refer to the other document.
+
+## Overview
+
+Firmware upgrades for LoRa Basics™ Station need to be supported in the CUPS
+protocol implementation. More information on the protocol and parameters
+exchanged between the Station and the CUPS server which are relevant to firmware
+upgrades can be found in [The CUPS protocol - documentation][cupsproto].
+
+## In-scope
+
+This document focuses on:
+
+- Defining a flow for executing firmware upgrades of the Basics Station
+- Defining the changes needed in device twins for the concentrator station
+  stored in IoT Hub for handling firmware upgrades
+- Defining the changes required in the storage solution used for the CUPS
+  protocol implementation
+- Defining the changes needed in the Azure Function for supporting firmware
+  upgrades
+- Defining the changes needed in LoRaWan Network Server (LNS) for handling
+  firmware upgrades
+- Defining the changes needed in LoRa Device Provisioning CLI for allowing
+  firmware upgrades
+
+## Out-of-scope
+
+- Generating of the signature key, CRC32 checksum of the signature and digest in
+  the LoRa Device Provisioning CLI is considered a stretch and will not be added
+  to the tool as a functionality for the time being. This document provides
+  sample commands which can be used to generate required values in the appendix.
+
+## Decisions
+
+### Context
+
+The CUPS request described in [The CUPS protocol - documentation][cupsproto]
+contains the `package` field which indicates the current firmware version of the
+Basics Station. The value will need to be updated whenever a firmware upgrade of
+the Basics Station is performed by the user.
+
+### Firmware upgrade flow
+
+1. When provisioning a device that will need firmware upgrades, the end user
+   should generate a `sig-0.key` and store it on the device and in a centralized
+   repository of their own choice (example on how to generate it can be found in
+   the appendix).
+
+1. When a firmware upgrade needs to be provided to the device, the user should
+   generate a digest of the executable file of the upgrade (as in examples
+   mentioned above) and retrieve the CRC32 Checksum of the `sig-0.key`. Then 4
+   inputs (file, digest, checksum and new version) should be provided to the
+   LoRa Device Provisioning CLI tool to properly upload the file to a blob
+   storage and store the needed information in the concentrator twin (blob URL
+   in storage, digest, checksum and the new version number).
+
+1. The Basics Station sends the CUPS request containing the currently used
+   version (in the `package` field).
+
+1. The LNS compares the values from the device twin and the CUPS request, and
+   determines if an upgrade is required. If that's the case, the LNS will
+   download the upgrade file from storage and send a properly populated response
+   to the Basics Station.
+
+1. The Basics Station will then execute the actual firmware upgrade.
```

Comment:
can we consider adding the reported package version? that will allow us to determine, when the concentrator actually updated.